### PR TITLE
Peer discovery review only [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by LeoHChen**
## Issue

#373 
This set of patches has enabled peer discovery in harmony using libp2p.

#### Test/Run Logs

run  test/test_peer_discovery.sh test/configs/pd_config1.txt

in the validator log files, you can find the following lines showing new peers added.

```
{"lvl":"info","msg":"AddPeer add to peerstore","peer":"127.0.0.1:9001/\u003cpeer.ID Qm*qKUg9t\u003e[1]","t":"2019-02-03T06:52:06.240526656Z"}
```

## TODO

- [ ] add gossiping support
- [ ] enable consensus
- [ ] get txgen/wallet working with bootnode and shard

### CODE REVIEW

File: cmd/harmony.go [hunk 1]
```diff
@@ -89,102 +86,86 @@ func main() {
...
+	// Key file to store the private key
+	keyFile := flag.String("key", "./.hmykey", "the private key file of the bootnode")
+	flag.Var(&utils.BootNodes, "bootnodes", "a list of bootnode multiaddress")
+
+	// The initial shard ID (TODO: leo) this is a workaround for initial blockchain bootstrap
+	shardID := flag.String("shardID", "0", "the shardID. 0 is beacon chain")
+	// This node is a leader node. (TODO: leo) this is also a workaround for initial blockchain bootstrap
+	isLeader := flag.Bool("leader", false, "false means the node is not a shard leader")
 
 	flag.Parse()
```
Comment (alajko) [inline cmd/harmony.go]: Is this specifically for beaconLeader (during bootstrap)?

Comment (LeoHChen) [reply cmd/harmony.go]: yes, I plan to use it that way.

File: api/service/discovery/service.go [hunk 2]
```diff
@@ -86,8 +90,11 @@ func (s *Service) foundPeers() {
 				p := p2p.Peer{PeerID: peer.ID, Addrs: peer.Addrs}
 				s.Host.AddPeer(&p)
 				// TODO: stop ping if pinged before
+				// TODO: call staking servcie here if it is a new node
 				s.pingPeer(p)
 			}
+		case <-s.stopChan:
```
Comment (harmony-ek) [inline api/service/discovery/service.go]: if we used a context, this would become `case <-ctx.Done():`

File: node/node.go [hunk 3]
```diff
@@ -674,9 +691,13 @@ func (node *Node) setupForShardLeader() {
 	node.serviceManager.RegisterService(service_manager.BlockProposal, blockproposal.NewService(node.Consensus.ReadySignal, node.WaitForConsensusReady))
 	// Register client support service.
 	node.serviceManager.RegisterService(service_manager.ClientSupport, clientsupport.NewService(node.blockchain.State, node.CallFaucetContract, node.SelfPeer.IP, node.SelfPeer.Port))
+	// Register peer discovery service.
+	node.serviceManager.RegisterService(service_manager.PeerDiscovery, discovery.New(node.host, fmt.Sprintf("%v", node.Consensus.ShardID)))
```
Comment (mikedoan) [inline node/node.go]: as discussed you said peerdiscovery does 2 things: 1/ for new node, it talks to bootnode to obtain info of some beacon peers
2/ maintain the discovery peers

i believe we need 2/ for every node.
1/ is just for new node or anyone who does not know how to join the blockchain network.

with that i'm thinking we can break down discovery into 2 different services.
what do you think?

File: node/node.go [hunk 4]
```diff
@@ -287,6 +296,14 @@ func New(host p2p.Host, consensus *bft.Consensus, db ethdb.Database) *Node {
 		}
 		node.Consensus.ConsensusBlock = make(chan *bft.BFTBlockInfo)
 		node.Consensus.VerifiedNewBlock = make(chan *types.Block)
+
+		var err error
+		node.Pubsub, err = pubsub.NewFloodSub(context.Background(), node.host.GetP2PHost())
```
Comment (mikedoan) [inline node/node.go]: pubsub - is it what eugene is working on now?

File: cmd/harmony.go [hunk 5]
```diff
@@ -210,13 +192,23 @@ func main() {
...
+		// go currentNode.JoinShard(leader)
 	}
 
 	go currentNode.SupportSyncing()
+	if consensus.IsLeader {
+		go currentNode.SupportClient()
+	}
+
+	log.Info("ServiceManagerSetup")
 	currentNode.ServiceManagerSetup()
```
Comment (mikedoan) [inline cmd/harmony.go]: sync with master again

File: node/node.go (hunk 3 above)
Comment (LeoHChen) [reply node/node.go]: it is okay to split.  So I will use networkinfo service to connect to bootnode and discovery service to connect to peers.

File: node/node.go (hunk 4 above)
Comment (LeoHChen) [reply node/node.go]: yes, this is a placeholder.

**Comment by LeoHChen** [comment]
this PR is no longer needed.

**Pull request closed without merging**